src/IO/HTTPCommon.cpp

```diff
@@ -105,23 +105,53 @@ namespace
         const std::string host;
         const UInt16 port;
         bool https;
+        const String proxy_host;
+        const UInt16 proxy_port;
+        bool proxy_https;
+        bool resolve_host;
         using Base = PoolBase<Poco::Net::HTTPClientSession>;
         ObjectPtr allocObject() override
         {
-            return makeHTTPSessionImpl(host, port, https, true);
+            auto session = makeHTTPSessionImpl(host, port, https, true, resolve_host);
+            if (!proxy_host.empty())
+            {
+                const String proxy_scheme = proxy_https ? "https" : "http";
+                session->setProxyHost(proxy_host);
+                session->setProxyPort(proxy_port);
+#ifdef POCO_CLICKHOUSE_PATCH
+                session->setProxyProtocol(proxy_scheme);
+                session->setProxyTunnel(false);
+#endif
+            }
+            return session;
         }
 
     public:
-        SingleEndpointHTTPSessionPool(const std::string & host_, UInt16 port_, bool https_, size_t max_pool_size_)
-            : Base(max_pool_size_, &Poco::Logger::get("HTTPSessionPool")), host(host_), port(port_), https(https_)
+        SingleEndpointHTTPSessionPool(
+                const std::string & host_,
+                UInt16 port_,
+                bool https_,
+                const std::string & proxy_host_,
+                UInt16 proxy_port_,
+                bool proxy_https_,
+                size_t max_pool_size_,
+                bool resolve_host_ = true)
+            : Base(max_pool_size_, &Poco::Logger::get("HTTPSessionPool"))
+            , host(host_)
+            , port(port_)
+            , https(https_)
+            , proxy_host(proxy_host_)
+            , proxy_port(proxy_port_)
+            , proxy_https(proxy_https_)
+            , resolve_host(resolve_host_)
         {
         }
     };
 
     class HTTPSessionPool : private boost::noncopyable
     {
     private:
-        using Key = std::tuple<std::string, UInt16, bool>;
+        using Key = std::tuple<std::string, UInt16, bool, std::string, UInt16, bool>;
```

Comment:
Which std::string is which? A struct would be more clear.

Reply:
+